src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineComputerLauncher.java

```diff
@@ -304,4 +320,97 @@ public void onClosed(Channel channel, IOException cause) {
+    final long startTime = System.currentTimeMillis();
+    while (true) {
+      try {
+        long waitTime = System.currentTimeMillis() - startTime;
+        if (timeout > 0 && waitTime > timeout) {
+          // TODO(google-compute-engine-plugin/issues/135): better exception
+          throw new Exception(
+              "Timed out after "
+                  + (waitTime / 1000)
+                  + " seconds of waiting for ssh to become available. (maximum timeout configured is "
+                  + (timeout / 1000)
+                  + ")");
+        }
+        Instance instance = computer.refreshInstance();
+
+        String host = "";
+
+        // TODO(google-compute-engine-plugin/issues/136): handle multiple NICs
+        NetworkInterface nic = instance.getNetworkInterfaces().get(0);
+
+        if (this.useInternalAddress) {
+          host = nic.getNetworkIP();
+        } else {
+          // Look for a public IP address
+          if (nic.getAccessConfigs() != null) {
+            for (AccessConfig ac : nic.getAccessConfigs()) {
+              if (ac.getType().equals(InstanceConfiguration.NAT_TYPE)) {
+                host = ac.getNatIP();
+              }
+            }
+          }
+          // No public address found. Fall back to internal address
+          if (host.isEmpty()) {
+            host = nic.getNetworkIP();
+          }
+        }
+
+        int port = SSH_PORT;
+        logInfo(
+            computer,
+            listener,
+            "Connecting to " + host + " on port " + port + ", with timeout " + SSH_TIMEOUT + ".");
+        Connection conn = new Connection(host, port);
+        ProxyConfiguration proxyConfig = Jenkins.get().proxy;
+        Proxy proxy = proxyConfig == null ? Proxy.NO_PROXY : proxyConfig.createProxy(host);
+        if (!node.isIgnoreProxy()
+            && !proxy.equals(Proxy.NO_PROXY)
+            && proxy.address() instanceof InetSocketAddress) {
+          InetSocketAddress address = (InetSocketAddress) proxy.address();
+          HTTPProxyData proxyData = null;
+          if (proxyConfig.getUserName() != null) {
+            proxyData =
+                new HTTPProxyData(
+                    address.getHostName(),
+                    address.getPort(),
+                    proxyConfig.getUserName(),
+                    proxyConfig.getPassword());
+          } else {
+            proxyData = new HTTPProxyData(address.getHostName(), address.getPort());
+          }
+          conn.setProxyData(proxyData);
+          logInfo(computer, listener, "Using HTTP Proxy Configuration");
+        }
+        // TODO(google-compute-engine-plugin/issues/137): verify host key
+        conn.connect(
+            new ServerHostKeyVerifier() {
+              public boolean verifyServerHostKey(
+                  String hostname, int port, String serverHostKeyAlgorithm, byte[] serverHostKey)
+                  throws Exception {
+                return true;
```

Comment:
I realize this is probably carried over code, but it caught my attention so....is there a enough data from the objects we're getting from the GCE API to provide actual host key verification?

Reply:
I've been looking into this and it seems like for the most part this is a common pattern across other Jenkins Cloud implementations for the launch/connect process. The ec2-plugin has an [implementation](https://github.com/jenkinsci/ec2-plugin/blob/597af44906321da78a92fd355355b77f5751a679/src/main/java/hudson/plugins/ec2/ssh/HostKeyVerifierImpl.java) which checks prior console logs for the fingerprint, but I couldn't find any usages of it within the plugin, and in the actual [EC2UnixLauncher](https://github.com/jenkinsci/ec2-plugin/blob/597af44906321da78a92fd355355b77f5751a679/src/main/java/hudson/plugins/ec2/ssh/EC2UnixLauncher.java#L406) they have this return true as well. I can add what I've found to the issue (#137) as a possible way to implement this in the future, but since this is just carried over I think it's out of scope for this PR.

Reply:
Thanks for looking into it. Could you add a comment explanation and a TODO for a later PR.